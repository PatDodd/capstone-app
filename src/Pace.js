import React, { Component, Fragment } from 'react';

import { CircularProgress, FormControl, InputLabel, Select, MenuItem, 
    Button, IconButton, Snackbar } from '@material-ui/core';
import { Close } from '@material-ui/icons/';
import { withStyles } from '@material-ui/core/styles';
import firebase from './Firebase';

import OneMileLine from './OneMileLine';

const styles = theme => ({
    formControl: {
        margin: theme.spacing(3),
        width: '10%',
        [theme.breakpoints.up('lg')]: {
            width: '10%',
        },
        [theme.breakpoints.down('md')]: {
            width: '15%',
        },
        [theme.breakpoints.down('sm')]: {
            width: '20%',
        },
        [theme.breakpoints.down('xs')]: {
            width: '45%',
        }
    },
    button: {
        color: 'white',
        backgroundColor: theme.palette.primary.light,
        '&:hover': {
            backgroundColor: theme.palette.primary.contrastText
        },
        padding: '10px'
    },
    ErrorBtn: {
        color: theme.palette.error.main
    }
})



class Pace extends Component {
    
    constructor(props){
        
        super(props);
        
        this.state = {
            data: [],
            options: [],
            distance: '',
            effort: "",
            from: "",
            to: "",
            ready: false,
            graphLoaded: false,
            error: false
        }

        // get ref to db
        const db = firebase.firestore();
    
        var arr = []; // to hold initial data from db

        // call db if there's nothing in data[] yet
        if(this.state.data.length === 0) {
            db.collection('runs').get().then((querySnapshot) =>{
                querySnapshot.forEach((doc) => {
                    arr.push(doc.data());
                });
            }).then(() => {
               
                this.setState({data: arr})
                const arr2 = this.state.data.map((item, i) => item.activityDate);
                this.setState({options: arr2}); 
            });
        }
         
        //console.log(this.state.data);
        this.handleChange = this.handleChange.bind(this);
        this.handleClick = this.handleClick.bind(this); 
    }//end CONSTRUCTOR

    

    handleChange(e){
        e.preventDefault();

        //toggle state to prevent change until click event occurs
        if(this.state.ready){
            this.setState({ready: false});
        } 
        if(e.target.name === "mileage"){
            e.preventDefault();
            const val = parseFloat(e.target.value);
            this.setState({ distance: val });
        }

        if(e.target.name === "effort"){ this.setState({ effort: e.target.value }); }

        var from;
        var to;
        if(e.target.name === "from"){
            from = e.target.value;
            this.setState({from: from});
        }

        if(e.target.name === "to"){
            to = e.target.value;
            this.setState({to: to});
        }

        
    }

    handleClick(e){
        e.preventDefault();
        let a = this.state.distance;
        let b = this.state.effort;
        let c = this.state.from;
        let g = this.state.to;
      
        if((a && b && c && g) && (new Date(c) <= new Date(g))){
            this.setState({ ready: true });
            if((a === "undefined" || b === "undefined")){
                //alert("form is incomplete");
                this.setState({error: true});
            }
            
        } else {
            //alert('form is incomplete or date span invalid');
            this.setState({error: true});
        }       
    }

    handleClose = (event, reason) => {
        if (reason === 'clickaway') {
          return;
        }
    
        this.setState({error: false});
    };

    render(){

        //get options from state
        const { options } = this.state;
        // classes for styling
        const { classes } = this.props;
        // parse options and format data as a string to insert into select box
        var opt = options.map((item, i) => {
            let d = new Date(item);
            let date = d.getDate();
            let month = d.getMonth()+1; // 0 indexed for some reason
            let year = d.getFullYear();
            let dateStr = `${month}/${date}/${year}`;
            return <MenuItem value={dateStr} key={i}>{dateStr}</MenuItem>;
        });
    
         // render bar graph when ready is true
        let chartComponent;

        if(this.state.ready){
            chartComponent = <OneMileLine 
                                data={this.state.data} 
                                distance={this.state.distance} 
                                effort={this.state.effort}
                                from={this.state.from}
                                to={this.state.to}
                             />
            
        }

        return(

            <Fragment>
                {opt.length === 0 && (
                    <CircularProgress color="primary"/>
                )}
                {opt !==0 && (
                <Fragment>
                    <FormControl className={classes.formControl}>
                        <InputLabel >Mileage: </InputLabel>
                        <Select name="mileage" onChange={this.handleChange} value={this.state.distance}>
                            <MenuItem value=""><em>None</em></MenuItem>
                            <MenuItem value="1">1</MenuItem>
                            <MenuItem value="2">2</MenuItem>
                            <MenuItem value="3">3</MenuItem>
                            <MenuItem value="4">4</MenuItem>
                            <MenuItem value="5">5</MenuItem>
                        </Select>
                    </FormControl>
                    <FormControl className={classes.formControl}>
                        <InputLabel >Effort Level: </InputLabel>
                        <Select name="effort" onChange={this.handleChange} value={this.state.effort}>
                            <MenuItem value=""><em>None</em></MenuItem>
                            <MenuItem value="all">All</MenuItem>
                            <MenuItem value="easy">Easy</MenuItem>
                            <MenuItem value="medium">Medium</MenuItem>
                            <MenuItem value="hard">Hard</MenuItem>
                        </Select>
                    </FormControl>
                    <FormControl className={classes.formControl}>
                        <InputLabel >From: </InputLabel>
                        <Select name="from" onChange={this.handleChange} value={this.state.from}>
                            <MenuItem value=""><em>None</em></MenuItem>
                            {opt}
                        </Select>
                    </FormControl>
                    <FormControl className={classes.formControl}>
                        <InputLabel>To: </InputLabel>
                        <Select name="to" onChange={this.handleChange} value={this.state.to}>
                            <MenuItem value=""><em>None</em></MenuItem>
                            {opt}
                        </Select>
                    </FormControl>
                <   FormControl className={classes.formControl}>
                        <Button type="submit" onClick={this.handleClick} className={classes.button}>Get Chart</Button>
                    </FormControl>
                    {chartComponent}
                </Fragment>
            )}
            <Snackbar
                anchorOrigin={{
                    vertical: 'top',
                    horizontal: 'left',
                    }}
                open={this.state.error}
                autoHideDuration={6000}
                onClose={this.handleClose}
                message="Form is incomplete or date range is invalid."
                action={
            <Fragment>
               
                <Button className={classes.ErrorBtn} size="small" onClick={this.handleClose}>
                ERROR
                </Button>
                <IconButton size="small" aria-label="close" color="inherit" onClick={this.handleClose}>
                <Close fontSize="small" />
                </IconButton>
            </Fragment>
        }
            /> 
        </Fragment>
        );
    }
}



export default withStyles(styles)(Pace);