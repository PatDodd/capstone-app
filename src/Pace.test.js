import React from 'react';
import ReactDOM from 'react-dom';
import Pace from './Pace';
import renderer from 'react-test-renderer';

// smoke test
it('renders <Pace /> component without crashing, smoke test', () => {
  const div = document.createElement('div'); 
  ReactDOM.render(<Pace />, div);
});

// snapshot test
it('renders correctly on snapshot test', () => {
  const tree = renderer.create(<Pace />).toJSON();
  expect(tree).toMatchSnapshot();
});