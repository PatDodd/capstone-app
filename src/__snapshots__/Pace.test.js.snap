// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly on snapshot test 1`] = `
Array [
  <div
    className="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
    role="progressbar"
    style={
      Object {
        "height": 40,
        "width": 40,
      }
    }
  >
    <svg
      className="MuiCircularProgress-svg"
      viewBox="22 22 44 44"
    >
      <circle
        className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
        cx={44}
        cy={44}
        fill="none"
        r={20.2}
        strokeWidth={3.6}
        style={Object {}}
      />
    </svg>
  </div>,
  <div
    className="MuiFormControl-root Pace-formControl-1"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
      data-shrink={false}
    >
      Mileage: 
    </label>
    <div
      className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
      onClick={[Function]}
    >
      <div
        aria-haspopup="listbox"
        aria-labelledby=" mui-component-select-mileage"
        className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
        id="mui-component-select-mileage"
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        role="button"
        tabIndex={0}
      >
        <span
          dangerouslySetInnerHTML={
            Object {
              "__html": "&#8203;",
            }
          }
        />
      </div>
      <input
        name="mileage"
        onAnimationStart={[Function]}
        type="hidden"
        value=""
      />
      <svg
        aria-hidden="true"
        className="MuiSvgIcon-root MuiSelect-icon"
        focusable="false"
        role="presentation"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
    </div>
  </div>,
  <div
    className="MuiFormControl-root Pace-formControl-1"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
      data-shrink={false}
    >
      Effort Level: 
    </label>
    <div
      className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
      onClick={[Function]}
    >
      <div
        aria-haspopup="listbox"
        aria-labelledby=" mui-component-select-effort"
        className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
        id="mui-component-select-effort"
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        role="button"
        tabIndex={0}
      >
        <span
          dangerouslySetInnerHTML={
            Object {
              "__html": "&#8203;",
            }
          }
        />
      </div>
      <input
        name="effort"
        onAnimationStart={[Function]}
        type="hidden"
        value=""
      />
      <svg
        aria-hidden="true"
        className="MuiSvgIcon-root MuiSelect-icon"
        focusable="false"
        role="presentation"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
    </div>
  </div>,
  <div
    className="MuiFormControl-root Pace-formControl-1"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
      data-shrink={false}
    >
      From: 
    </label>
    <div
      className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
      onClick={[Function]}
    >
      <div
        aria-haspopup="listbox"
        aria-labelledby=" mui-component-select-from"
        className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
        id="mui-component-select-from"
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        role="button"
        tabIndex={0}
      >
        <span
          dangerouslySetInnerHTML={
            Object {
              "__html": "&#8203;",
            }
          }
        />
      </div>
      <input
        name="from"
        onAnimationStart={[Function]}
        type="hidden"
        value=""
      />
      <svg
        aria-hidden="true"
        className="MuiSvgIcon-root MuiSelect-icon"
        focusable="false"
        role="presentation"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
    </div>
  </div>,
  <div
    className="MuiFormControl-root Pace-formControl-1"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
      data-shrink={false}
    >
      To: 
    </label>
    <div
      className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
      onClick={[Function]}
    >
      <div
        aria-haspopup="listbox"
        aria-labelledby=" mui-component-select-to"
        className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
        id="mui-component-select-to"
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        role="button"
        tabIndex={0}
      >
        <span
          dangerouslySetInnerHTML={
            Object {
              "__html": "&#8203;",
            }
          }
        />
      </div>
      <input
        name="to"
        onAnimationStart={[Function]}
        type="hidden"
        value=""
      />
      <svg
        aria-hidden="true"
        className="MuiSvgIcon-root MuiSelect-icon"
        focusable="false"
        role="presentation"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
    </div>
  </div>,
  <div
    className="MuiFormControl-root Pace-formControl-1"
  >
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-text Pace-button-2"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="submit"
    >
      <span
        className="MuiButton-label"
      >
        Get Chart
      </span>
    </button>
  </div>,
]
`;
